File: api/src/functions/eventPurchased.ts
import { app, HttpRequest, HttpResponseInit, InvocationContext } from '@azure/functions'
import { getById, update } from '../cosmosdb.js'
import { getTroopContext, unauthorized, forbidden } from '../middleware/auth.js'
import { checkPermission } from '../middleware/roles.js'
import { togglePurchasedItemSchema, validationError } from '../schemas.js'

const CONTAINER = 'events'

async function eventPurchasedHandler(req: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
  const id = req.params.id
  context.log(`PATCH /api/events/${id}/purchased`)

  const auth = await getTroopContext(req, context)
  if (!auth) return unauthorized()
  if (!checkPermission(auth.role, 'viewContent')) return forbidden()

  try {
    const parsed = togglePurchasedItemSchema.safeParse(await req.json())
    if (!parsed.success) return validationError(parsed.error)

    const existing = await getById(CONTAINER, id, auth.troopId)
    if (!existing) return { status: 404, jsonBody: { error: 'Event not found' } }

    const purchasedItems = new Set<string>(existing.purchasedItems ?? [])
    if (parsed.data.purchased) {
      purchasedItems.add(parsed.data.item)
    } else {
      purchasedItems.delete(parsed.data.item)
    }

    const updatedEvent = await update(
      CONTAINER,
      id,
      {
        ...existing,
        id,
        troopId: auth.troopId,
        purchasedItems: Array.from(purchasedItems),
        updatedAt: Date.now(),
        updatedBy: { userId: auth.userId, displayName: auth.displayName },
      },
      auth.troopId
    )

    return { jsonBody: updatedEvent }
  } catch (err) {
    context.error(`PATCH /api/events/${id}/purchased failed:`, err)
    return { status: 500, jsonBody: { error: 'Internal server error' } }
  }
}

app.http('eventPurchased', {
  methods: ['PATCH'],
  authLevel: 'anonymous',
  route: 'events/{id}/purchased',
  handler: eventPurchasedHandler,
})

File: api/app/routers/event_purchased.py
from __future__ import annotations

import time

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from app.cosmosdb import get_by_id, update_item
from app.middleware.auth import RequireTroopContext, forbidden
from app.middleware.roles import check_permission
from app.schemas import TogglePurchasedItem

router = APIRouter()

CONTAINER = "events"


@router.patch("/events/{event_id}/purchased")
async def toggle_purchased(event_id: str, body: TogglePurchasedItem, auth: RequireTroopContext):
    if not check_permission(auth.role, "viewContent"):
        forbidden()

    existing = await get_by_id(CONTAINER, event_id, auth.troopId)
    if not existing:
        return JSONResponse({"error": "Event not found"}, status_code=404)

    purchased_items = set(existing.get("purchasedItems") or [])
    if body.purchased:
        purchased_items.add(body.item)
    else:
        purchased_items.discard(body.item)

    updated = await update_item(
        CONTAINER,
        event_id,
        {
            **existing,
            "id": event_id,
            "troopId": auth.troopId,
            "purchasedItems": list(purchased_items),
            "updatedAt": int(time.time() * 1000),
            "updatedBy": {"userId": auth.userId, "displayName": auth.displayName},
        },
        auth.troopId,
    )
    return updated
